arangod/VocBase/LogicalDataSource.h

```diff
@@ -57,16 +59,28 @@ class LogicalDataSource {
   virtual ~LogicalDataSource() = default;
 
   Category category() const noexcept { return _category; }
+
   [[nodiscard]] bool deleted() const noexcept {
     return _deleted.load(std::memory_order_relaxed);
   }
+
   void deleted(bool deleted) noexcept {
-    _deleted.store(deleted, std::memory_order_relaxed);
+    TRI_ASSERT(deleted);
+    // intentionally do not use memory_order_relaxed here, as
+    // the deleted(true) call will only happen once in the lifetime
+    // of a data source, and it is not necessary to optimize the
+    // performance of it.
+    _deleted.store(deleted);
   }
```

Comment:
```suggestion
  void setDeleted() noexcept {
    #ifdef ARANGODB_MAINTAINER_MODE
      TRI_ASSERT(!_deleted.exchange(true));
    #else
      // relaxed here and in load ok because we don't need
      // happens before between them.
      _deleted.store(true, std::memory_order_relaxed);
    #endif
}
```